### PR TITLE
issue#10 | Estimate user's next purchase

### LINKED ISSUE #10: 10. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items.
AC:

* When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

The script at `/src/lib/estimates.js` should be used to calculate the next estimated purchase interval.

### PR DESCRIPTION
**Pull request by yahayakenny**
## Description

This PR adds two new attributes to the item object, the `numberOfPurchases`, and the `previousPurchase`, It estimates how soon the user is going to buy the items and stores the dates as integers instead of strings.

## Related Issue

Closes #10

## Acceptance Criteria

- [x] When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Testing Steps / QA Criteria
* From your terminal, pull down this branch with `git pull origin yk-se-issue10` and checkout that branch with `git checkout yk-se-issue10 ` 
* Run `npm start` to launch the app
* In the app, create a new list and add a few items
* Go to the List view and check one of the newly added list items. 
* Open up the [Firestore console](https://console.firebase.google.com/u/1/project/tcl-15-smart-shopping-list/firestore/data~2FshoppingLists~2FZc4ke3K09gJkaWmXE7uC) and take a look at the checked item, you should notice that previousPurchase and recentPurchase for the unchecked items are initialized to `null` and numberOfPuchases is 0. However,  the recentPurchase for the checked item has a value and the numberOfPurchases is 1.
* Now add a value less than the recent purchase to simulate that more than 24hrs has passed since the last purchase.
* Return to the app, you will notice the item is now unchecked.
* Check it a second time, return to the console and observe the value of `howSoon` update to reflect the estimated days till next purchase and the numberOfPurchases increase by 1.

### CODE REVIEW

File: src/components/ListItem.js
Comment (hajarNasr) [inline src/components/ListItem.js]: I'm too sleepy right now, so, sorry if this sounds stupid or anything but why are you using `+=` here? 😅

File: src/components/ListItem.js
Comment (hajarNasr) [inline src/components/ListItem.js:25]: Hey, Sarah or Kenny! Could you be so kind and write me a little explanation on `calculateEstimate()` here? I'm not sure if I understand it correctly. 😅

File: src/lib/helper.js
Comment (hajarNasr) [inline src/lib/helper.js:39]: Love the names of the functions.

File: src/lib/helper.js
Comment (hajarNasr) [inline src/lib/helper.js:27]: Thank you for writing the comments here! ❤️

**Review by hajarNasr — COMMENTED** [review]
Everything is looking awesome here and you two did a great job! 👏 ❤️

File: src/components/ListItem.js
Comment (saraheisa) [reply src/components/ListItem.js]: OH! It's a pure mistake that we didn't even notice 😅
Thanks for pointing it out

File: src/components/ListItem.js
Comment (saraheisa) [reply src/components/ListItem.js:25]: It's an algorithm🤣

File: src/lib/helper.js
Comment (deboragaleano) [reply src/lib/helper.js:27]: Nice ! 👍

File: src/lib/helper.js
Comment (deboragaleano) [reply src/lib/helper.js:39]: Agreed!

**Review by deboragaleano — COMMENTED** [review]
I followed the testing steps and it worked as described! I checked the item twice and the `howSoon` attribute was indeed updated (more than the value `7`) and the `numberOfPurchases` value increased by 1. 
Also, the item was unchecked when decreasing the value of the `recentPurchase` attribute ! 👍

File: src/components/AddItem.js
Comment (segdeha) [inline src/components/AddItem.js:33]: How are you using `previousPurchase` and `recentPurchase`? You can satisfy the AC by having only 1 data point that represents the date of the most recent purchase, but maybe you have another reason for keeping the other date around?

File: src/components/ListItem.js
Comment (segdeha) [reply src/components/ListItem.js:25]: We can discuss it more on the call, but it essentially is a weighted average where the current interval between the last and more recent purchases counts a little more than the previously calculated estimated number of days until the next purchase.

File: src/lib/helper.js
Comment (segdeha) [inline src/lib/helper.js:33]: I’m curious what made you choose `floor` and not `round` or `ceil`?

File: src/lib/helper.js
Comment (segdeha) [reply src/lib/helper.js:39]: Agreed, very descriptive!

**Review by segdeha — APPROVED** [review]
Mostly questions and comments from me. Didn’t see anything blocking and it all worked great in my testing!

I’m setting it to approved, but you should wait for peer feedback before merging. Nice work!

File: src/components/AddItem.js
Comment (Lukabaramishvili) [reply src/components/AddItem.js:33]: I'm curious about this as well. I also think that only `previousPurchase` should do it.

File: src/components/ListItem.js
Comment (Lukabaramishvili) [inline src/components/ListItem.js:16]: Maybe having the `previousPurchase` property as timestamp, it would be easier to test new features of the estimation?

**Review by Lukabaramishvili — APPROVED** [review]
Everything works as expected. Just left some comments and questions. Thanks! 👍

File: src/lib/helper.js
Comment (asianvader) [reply src/lib/helper.js:27]: I like that these date related functions are extracted out. Great in case you need to re-use it elsewhere.

**Review by asianvader — APPROVED** [review]
Followed your instructions and managed to get this working locally. Looks great so far and everything is working as expected. Nothing blocking, just comments.

File: src/lib/helper.js
Comment (yahayakenny) [reply src/lib/helper.js:33]: Well no particular reason tbh, I guess we just used the method that came to mind most readily😅

File: src/components/AddItem.js
Comment (saraheisa) [reply src/components/AddItem.js:33]: We added `previousPurchase` to test `calculateEstimate` 
but it's true we don't need to store it in DB
will remove it now:))

File: src/lib/helper.js
Comment (saraheisa) [reply src/lib/helper.js:33]: After some discussions, we decided that `Math.ceil` makes more sense
Because if we supposed it's the first day in the world of JS and we passed 0.3 of it 
then using `floor` and `round` will make it day 0 and using `ceil` will make it day 1
and I think it should be counted as day 1
Thanks a lot for the valuable comments Andrew 💜

**Pull request merged**
